source/extensions/filters/http/header_mutation/header_mutation.cc

```diff
@@ -0,0 +1,77 @@
+#include "source/extensions/filters/http/header_mutation/header_mutation.h"
+
+#include <cstdint>
+#include <memory>
+
+#include "source/common/config/utility.h"
+#include "source/common/http/header_map_impl.h"
+#include "source/common/http/utility.h"
+
+namespace Envoy {
+namespace Extensions {
+namespace HttpFilters {
+namespace HeaderMutation {
+
+void Mutations::mutateRequestHeaders(Http::RequestHeaderMap& request_headers,
+                                     const StreamInfo::StreamInfo& stream_info) const {
+  request_mutations_.evaluateHeaders(request_headers, request_headers,
+                                     *Http::StaticEmptyHeaders::get().response_headers,
+                                     stream_info);
+}
+
+void Mutations::mutateResponseHeaders(const Http::RequestHeaderMap& request_headers,
+                                      Http::ResponseHeaderMap& response_headers,
+                                      const StreamInfo::StreamInfo& stream_info) const {
+  response_mutations_.evaluateHeaders(response_headers, request_headers, response_headers,
+                                      stream_info);
+}
+
+PerRouteHeaderMutation::PerRouteHeaderMutation(const PerRouteProtoConfig& config)
+    : mutations_(config.mutations()) {}
+
+HeaderMutationConfig::HeaderMutationConfig(const ProtoConfig& config)
+    : mutations_(config.mutations()) {}
+
+Http::FilterHeadersStatus HeaderMutation::decodeHeaders(Http::RequestHeaderMap& headers, bool) {
+  config_->mutations().mutateRequestHeaders(headers, decoder_callbacks_->streamInfo());
+
+  // Only the most specific route config is used.
+  // TODO(wbpcode): It's possible to traverse all the route configs to merge the header mutations
+  // in the future.
+  route_config_ =
+      Http::Utility::resolveMostSpecificPerFilterConfig<PerRouteHeaderMutation>(decoder_callbacks_);
+
+  if (route_config_ != nullptr) {
+    route_config_->mutations().mutateRequestHeaders(headers, decoder_callbacks_->streamInfo());
+  }
+
+  return Http::FilterHeadersStatus::Continue;
+}
+
+Http::FilterHeadersStatus HeaderMutation::encodeHeaders(Http::ResponseHeaderMap& headers, bool) {
+  const Http::RequestHeaderMap& request_headers =
+      encoder_callbacks_->streamInfo().getRequestHeaders() == nullptr
+          ? *Http::StaticEmptyHeaders::get().request_headers
```

Comment:
Why is this empty? Don't we know the request headers from earlier?

Reply:
In fact, logically to say, the request headers will never be nullptr.

I will update these code to use a ASSERT here to avoid the nullptr check.